drivers/volume/portworx/portworx.go

```diff
@@ -388,6 +629,295 @@ func (p *portworx) VolumeDelete(pv *v1.PersistentVolume) error {
+	if err != nil {
+		return nil, err
+	}
+
+	snapDataSource.PortworxSnapshot.SnapshotData = snapData.Metadata.Name
+
+	snap := &crdv1.VolumeSnapshot{
+		Metadata: metav1.ObjectMeta{
+			Name:      volumeSnapshotName,
+			Namespace: namespace,
+			Labels:    snapLabels,
+		},
+		Spec: crdv1.VolumeSnapshotSpec{
+			SnapshotDataName:          snapData.Metadata.Name,
+			PersistentVolumeClaimName: parentPVC.Name,
+		},
+		Status: crdv1.VolumeSnapshotStatus{
+			Conditions: snapStatus,
+		},
+	}
+
+	logrus.Infof("Creating VolumeSnapshot object [%#v]", snap)
+	err = wait.ExponentialBackoff(snapAPICallBackoff, func() (bool, error) {
+		_, err := k8s.Instance().CreateSnapshot(snap)
+		if err != nil {
+			logrus.Errorf("failed to create volumesnapshot due to err: %v", err)
+			return false, nil
+		}
+
+		return true, nil
+	})
+	return snap, err
+}
+
+func (p *portworx) createVolumeSnapshotData(
+	snapshotName, snapshotNamespace string,
+	snapDataSource *crdv1.VolumeSnapshotDataSource,
+	snapStatus *[]crdv1.VolumeSnapshotCondition,
+	snapLabels map[string]string) (*crdv1.VolumeSnapshotData, error) {
+	var lastCondition crdv1.VolumeSnapshotDataCondition
+	if snapStatus != nil && len(*snapStatus) > 0 {
+		conditions := *snapStatus
+		ind := len(conditions) - 1
+		lastCondition = crdv1.VolumeSnapshotDataCondition{
+			Type:    (crdv1.VolumeSnapshotDataConditionType)(conditions[ind].Type),
+			Status:  conditions[ind].Status,
+			Message: conditions[ind].Message,
+		}
+	}
+
+	snapshotData := &crdv1.VolumeSnapshotData{
+		Metadata: metav1.ObjectMeta{
+			Name:   snapshotName,
+			Labels: snapLabels,
+		},
+		Spec: crdv1.VolumeSnapshotDataSpec{
+			VolumeSnapshotRef: &v1.ObjectReference{
+				Kind:      "VolumeSnapshot",
+				Name:      snapshotName,
+				Namespace: snapshotNamespace,
+			},
+			VolumeSnapshotDataSource: *snapDataSource,
+		},
+		Status: crdv1.VolumeSnapshotDataStatus{
+			Conditions: []crdv1.VolumeSnapshotDataCondition{
+				lastCondition,
+			},
+		},
+	}
+
+	var result *crdv1.VolumeSnapshotData
+	err := wait.ExponentialBackoff(snapAPICallBackoff, func() (bool, error) {
+		var err error
+		result, err = k8s.Instance().CreateSnapshotData(snapshotData)
+		if err != nil {
+			logrus.Errorf("failed to create snapshotdata due to err: %v", err)
+			return false, nil
+		}
+
+		return true, nil
+	})
+
+	if err != nil {
+		logrus.Errorf("Error creating the VolumeSnapshotData for snap %s: %v", snapshotName, err)
+		return nil, fmt.Errorf("Failed to create the VolumeSnapshotData for snap %s", snapshotName)
+	}
+
+	return result, nil
+}
+
+func (p *portworx) findParentPVCBySnapID(snapID string) (*v1.PersistentVolumeClaim, error) {
+	pxSnaps, err := p.volDriver.Inspect([]string{snapID})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(pxSnaps) != 1 {
+		return nil, fmt.Errorf("failed to inspect Portworx snapshot: %s. Received: %d entries in response",
+			snapID, len(pxSnaps))
+	}
+
+	pxSnap := pxSnaps[0]
+
+	if len(pxSnap.Source.Parent) == 0 {
```

Comment:
Check for "pxSnap.Source" not being nil before accessing. We were returning nil in some cases, which will cause a panic here.